### ISSUE TITLE
cache sparse (zero) chunks

### ISSUE THREAD

**Issue by enkore** [borgbackup/borg#1678]
When chunking sparse files the chunker will converge on an "idle tone" for runs of zeroes ~>= 2 chunks.

When extracting these are fetched over-and-over again, and also decrypted, checked etc. making it more slow than it has to be.

Suggestion:

LRUCache (chunk-id,) -> (length,) whose express purpose is to store all-zero chunks when --sparse is used. This needs a bit of work in extract_file and in the DownloadPipeline. As usual preload_ids may make this harder to implement (therefore creating this issue, so this doesn't get buried in my stack of notes). If we figure out #1665 this shouldn't be hard then - basically the same problem description regarding preload.

**Comment by ThomasWaldmann** [comment]
Yes, that is not very efficient - but is it really a problem? With compression, that all-zero chunk should be tiny.

**Comment by enkore** [comment]
It should. We still do the ID check though, limiting speed to something like 300-400 MB/s instead of ~infty.

**Comment by ThomasWaldmann** [comment]
See also #1354 and #14.

**Comment by ThomasWaldmann** [comment]
note: `borg create` now uses such a `LRUcache` and maps `(hashalgo, size) -> hashvalue`.

I decided to add the hashalgo to play a bit safer although currently it likely is always the same for a normal borg cli invocation (but maybe not for unit tests or future use).

**Comment by ThomasWaldmann** [comment]
Note: for `RemoteRepository`, borg uses a `RepositoryCache`.
